crates/polars-core/src/chunked_array/ops/gather.rs

```diff
@@ -0,0 +1,167 @@
+use arrow::array::Array;
+use polars_error::{polars_bail, polars_ensure, PolarsResult};
+
+use crate::chunked_array::ops::{ChunkTake, ChunkTakeUnchecked};
+use crate::chunked_array::ChunkedArray;
+use crate::datatypes::{IdxCa, PolarsDataType, StaticArray};
+use crate::prelude::*;
+use crate::utils::index_to_chunked_index;
+
+impl<T: PolarsDataType, I: AsRef<[IdxSize]> + ?Sized> ChunkTake<I> for ChunkedArray<T>
+where
+    ChunkedArray<T>: ChunkTakeUnchecked<I>,
+{
+    /// Gather values from ChunkedArray by index.
+    fn take(&self, indices: &I) -> PolarsResult<Self> {
+        let len = self.len();
+        let all_valid = indices.as_ref().iter().all(|i| (*i as usize) < len);
+        polars_ensure!(all_valid, ComputeError: "invalid index in gather");
+
+        // SAFETY: we just checked the indices are valid.
+        Ok(unsafe { self.take_unchecked(indices) })
+    }
+}
+
+unsafe fn target_value_unchecked<'a, A: StaticArray>(
+    targets: &[&'a A],
+    idx: IdxSize,
+) -> A::ValueT<'a> {
+    let (chunk_idx, arr_idx) =
+        index_to_chunked_index(targets.iter().map(|a| a.len()), idx as usize);
+    let arr = targets.get_unchecked(chunk_idx);
+    arr.value_unchecked(arr_idx)
+}
+
+unsafe fn target_get_unchecked<'a, A: StaticArray>(
+    targets: &[&'a A],
+    idx: IdxSize,
+) -> Option<A::ValueT<'a>> {
+    let (chunk_idx, arr_idx) =
+        index_to_chunked_index(targets.iter().map(|a| a.len()), idx as usize);
+    let arr = targets.get_unchecked(chunk_idx);
+    arr.get_unchecked(arr_idx)
+}
+
+unsafe fn gather_idx_array_unchecked<A: StaticArray>(
+    dtype: DataType,
+    targets: &[&A],
+    has_nulls: bool,
+    indices: &[IdxSize],
+) -> A {
+    let it = indices.iter().copied();
+    if targets.len() == 1 {
+        let arr = targets.iter().next().unwrap();
+        if has_nulls {
+            it.map(|i| arr.get_unchecked(i as usize))
+                .collect_arr_with_dtype(dtype)
+        } else {
+            it.map(|i| arr.value_unchecked(i as usize))
+                .collect_arr_with_dtype(dtype)
+        }
+    } else {
```

Comment:
I think we should do a benchmark and come up with a certain size of chunks at which we simply rechunk before we gather. Computing the offsets and having the double indirection is super expensive.

Reply:
I don't think the double indirection is particularly expensive, since you will always hit L1 cache on the first indirection assuming the number of chunks is reasonable (say <= 256). It's the calculating of offsets that's the expensive bit.

I had the idea of perhaps doing that with a branchless binary search, always with a fixed size. For example we can support up to 16 arrays by precomputing a size 16 `boundary` array (where the first element is a dummy, never accessed) and doing

```rust
let mut chunk_idx = 0;
chunk_idx |= if boundary[chunk_idx | 0b1000] < idx { 0b1000 } else { 0 };
chunk_idx |= if boundary[chunk_idx | 0b0100] < idx { 0b0100 } else { 0 };
chunk_idx |= if boundary[chunk_idx | 0b0010] < idx { 0b0010 } else { 0 };
chunk_idx |= if boundary[chunk_idx | 0b0001] < idx { 0b0001 } else { 0 };
```

There's also the possibility of SIMD to compare 4-8 indices at once, but that requires at least AVX: https://youtu.be/1RIPMQQRBWk?t=2218.